src/lib.rs

```diff
@@ -136,6 +144,98 @@ where
     }
 }
 
+impl<P, T> Atom<P>
+where
+    P: IntoRawPtr + FromRawPtr + Deref<Target = T>,
+{
+    /// Stores a value into the pointer if the current value is the same as the
+    /// current value.
+    ///
+    /// The return value is always the previous value. If it is equal to
+    /// current, then the value was updated.
+    ///
+    /// compare_and_swap also takes an Ordering argument which describes the
+    /// memory ordering of this operation.
+    pub fn compare_and_swap(
+        &self,
+        current: Option<&P>,
+        new: Option<P>,
+        order: Ordering,
+    ) -> Result<Option<P>, (Option<P>, *mut P)> {
+        let pcurrent = Self::inner_as_ptr(current);
+        let pnew = Self::inner_into_raw(new);
+        let pprev = self.inner.compare_and_swap(pcurrent, pnew, order);
+        if pprev == pcurrent {
+            Ok(unsafe { Self::inner_from_raw(pprev) })
+        } else {
+            Err((unsafe { Self::inner_from_raw(pnew) }, pprev as *mut P))
+        }
+    }
+
+    /// Stores a value into the pointer if the current value is the same as the
+    /// `current` value.
+    ///
+    /// The return value is a result indicating whether the new value was
+    /// written and containing the previous value. On success this value is
+    /// guaranteed to be equal to `current`.
+    ///
+    /// `compare_exchange` takes two `Ordering` arguments to describe the
+    /// memory ordering of this operation. The first describes the required
+    /// ordering if the operation succeeds while the second describes the
+    /// required ordering when the operation fails. The failure ordering can't
+    /// be `Release` or `AcqRel` and must be equivalent or weaker than the
+    /// success ordering.
+    pub fn compare_exchange(
```

Comment:
What exactly is the difference between `compare_and_swap` and `compare_exchange` here?

Reply:
I think except for the fact that you can specify a second ordering there is none.

Reply:
That's correct. Here's the source in AtomicPtr: https://doc.rust-lang.org/src/core/sync/atomic.rs.html#818-823